File: lib/distable.ex
defmodule Distable do
  @moduledoc """
  This is the public Elixir API for `:distable`.
  """
  use Application

  def start(_type, _args) do
    Distable.Supervisor.start_link()
  end

  @doc """
  Registers a name to a process started by the provided
  module/function/args. If the MFA does not start a process,
  an error will be returned.
  """
  @spec register(term, atom(), atom(), [term]) :: {:ok, pid} | {:error, term}
  def register(name, module, function, args)
    when is_atom(module) and is_atom(function) and is_list(args) do
    Distable.Tracker.register(name, {module, function, args})
  end
  @doc """
  Unregisters a name.
  """
  @spec unregister(term) :: :ok
  def unregister(name) do
    Distable.Tracker.unregister(name)
  end
  @doc """
  Add metadata to a registered process.
  """
  @spec register_property(pid, term) :: :ok
  def register_property(pid, prop) do
    Distable.Tracker.register_property(pid, prop)
  end
  @doc """
  Get the pid of a registered name.
  """
  @spec whereis(term) :: pid | :undefined
  def whereis(name) do
    Distable.Tracker.whereis(name)
  end
  @doc """
  Get a list of pids which have the given property in their
  metadata.
  """
  @spec get_by_property(term) :: [pid]
  def get_by_property(prop) do
    Distable.Tracker.get_by_property(prop)
  end
  @doc """
  Call the server associated with a given name.
  Use like `GenServer.call/3`
  """
  @spec call(term, term) :: term | {:error, term}
  @spec call(term, term, pos_integer) :: term | {:error, term}
  def call(name, message, timeout \\ 5_000) do
    Distable.Tracker.call(name, message, timeout)
  end
  @doc """
  Cast a message to a server associated with the given name.
  Use like `GenServer.cast/2`
  """
  @spec cast(term, term) :: :ok
  def cast(name, message) do
    Distable.Tracker.cast(name, message)
  end
end

File: src/distable.erl
-module(distable).

-export([
  start/2, stop/1,
  register/4, register_property/2, unregister/1,
  whereis/1, get_by_property/1,
  call/2, call/3, cast/2
]).

-define(TRACKER, 'Elixir.Distable.Tracker').

%% @doc You shouldn't need this if you've added the
%% the `distable` application to your applications
%% list, but it's here if you need it.
%% @end
start(Type, Args) ->
    'Elixir.Distable':start(Type, Args).

%% Same as above, use if you need it.
stop(State) ->
    'Elixir.Distable':stop(State).

%% @doc Registers a name to a process started by the provided
%% module/function/args. If the MFA does not start a process,
%% an error will be returned.
%% @end
-spec register(term(), atom(), atom(), [term()]) -> {ok, pid()} | {error, term()}.
register(Name, Module, Function, Args) ->
    ?TRACKER:register(Name, {Module, Function, Args}).

%% @doc Unregisters a name.
-spec unregister(term()) -> ok.
unregister(Name) ->
    ?TRACKER:unregister(Name).

%% @doc Add metadata to a registered process.
-spec register_property(pid(), term()) -> ok.
register_property(Pid, Prop) ->
    ?TRACKER:register_property(Pid, Prop).

%% @doc Get the pid of a registered name.
-spec whereis(term()) -> pid() | undefined.
whereis(Name) ->
    ?TRACKER:whereis(Name).

%% @doc Get a list of pids which have the given property in their metadata.
-spec get_by_property(term()) -> [pid].
get_by_property(Prop) ->
    ?TRACKER:get_by_property(Prop).

%% @doc Call the server associated with a given name.
%% Use like `gen_server:call/3`
%% @end
-spec call(term(), term()) -> term() | {error, term()}.
call(Name, Message) ->
    call(Name, Message, 5000).

%% @doc Call the server associated with a given name.
%% Same as `call/2`, but takes a timeout value for the call.
%% Use like `gen_server:call/3`
%% @end
-spec call(term(), term(), pos_integer()) -> term() | {error, term()}.
call(Name, Message, Timeout) ->
    ?TRACKER:call(Name, Message, Timeout).

%% @doc Cast a message to a server associated with the given name.
%% Use like `gen_server:cast/2`
%% @end
-spec cast(term(), term()) -> ok.
cast(Name, Message) ->
    ?TRACKER:cast(Name, Message).
